src/App.tsx

```diff
@@ -1,26 +1,425 @@
+import { addTodo, deleteTodo, getTodos, USER_ID } from './api/todos';
+import { Todo } from './types/Todo';
+import classNames from 'classnames';
 
-const USER_ID = 0;
+enum Status {
+  all = 'all',
+  active = 'active',
+  completed = 'completed',
+}
+
+function getVisibleTodos(todos: Todo[], status: Status) {
+  const copyTodos = [...todos];
+
+  if (status === Status.active) {
+    return copyTodos.filter(todo => !todo.completed);
+  }
+
+  if (status === Status.completed) {
+    return copyTodos.filter(todo => todo.completed);
+  }
+
+  return copyTodos;
+}
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [errorMessage, setErrorMessage] = useState('');
+  const [status, setStatus] = useState<Status>(Status.all);
+  const [textField, setTextField] = useState('');
+  const [isSubmiting, setIsSubmiting] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [isLoading, setIsLoading] = useState(0);
+  const field = useRef<HTMLInputElement>(null);
+  const completedTodos = todos.filter(todo => todo.completed).map(el => el.id);
+
+  const visibleTodos = getVisibleTodos(todos, status);
+
+  useEffect(() => {
+    getTodos()
+      .then(setTodos)
+      .catch(() => setErrorMessage('Unable to load todos'));
+  }, []);
+
+  useEffect(() => {
+    field.current?.focus();
+  }, [isSubmiting]);
+
+  useEffect(() => {
+    window.setTimeout(() => setErrorMessage(''), 3000);
+  }, [errorMessage]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleTextField = (e: React.ChangeEvent<HTMLInputElement>) => {
+    setTextField(e.target.value);
+    setErrorMessage('');
+  };
+
+  const handleSubmit = (e: FormEvent) => {
+    e.preventDefault();
+
+    if (!textField.trim()) {
+      setErrorMessage('Title should not be empty');
+
+      return;
+    }
+
+    const newTodo: Omit<Todo, 'id'> = {
+      userId: USER_ID,
+      title: textField,
+      completed: false,
+    };
+
+    setIsSubmiting(true);
+    setTempTodo({
+      id: 0,
+      userId: USER_ID,
+      title: textField,
+      completed: false,
+    });
+
+    addTodo(newTodo)
+      .then(todo => {
+        setTodos(currentTodos => [...currentTodos, todo]);
+        setTextField('');
+      })
+      .catch(error => {
+        setErrorMessage('Unable to add a todo');
+        throw error;
+      })
+      .finally(() => {
+        setIsSubmiting(false);
+        setTempTodo(null);
+      });
+  };
+
+  const handleDelete = (todoId: number) => {
+    setIsLoading(todoId);
+    deleteTodo(todoId)
+      .then(() =>
+        setTodos(currentTodos =>
+          currentTodos.filter(todo => todo.id !== todoId),
+        ),
+      )
+      .catch(error => {
+        setTodos(todos);
+        setErrorMessage('Unable to delete a todo');
+        throw error;
+      })
+      .finally(() => setIsLoading(0));
+  };
+
+  const handleDeleteCompleted = () => {
```

Comment:
It's better to use Promise.allSettled here and proceed results

Reply:
> It's better to use Promise.allSettled here and proceed results

through the loop, push all promises into some array and then write Promise.allSettled(...array)?